libraries/ESP8266WiFi/src/ESP8266WiFiMulti.cpp

```diff
@@ -1,267 +1,507 @@
+}
+
+/**
+ * @brief Remove all Access Points from list
+ */
+void ESP8266WiFiMulti::cleanAPlist(void)
+{
+    APlistClean();
+}
+
+/**
+ * @brief Check if Access Point exists in list
+ * @param ssid
+ *      WiFi SSID
+ * @param passphrase
+ *      WiFi Password
+ * @retval true
+ *      Success
+ * @retval false
+ *      Failure
+ */
+bool ESP8266WiFiMulti::existsAP(const char *ssid, const char *passphrase)
+{
+    return APlistExists(ssid, passphrase);
+}
+
+/**
+ * @brief Keep WiFi connected to Access Point with strongest WiFi signal (RSSI)
+ * @param connectTimeoutMs
+ *      Timeout in ms per WiFi connection (excluding fixed 5 seconds scan timeout)
+ * @return
+ *      WiFi status
+ */
+wl_status_t ESP8266WiFiMulti::run(uint32_t connectTimeoutMs)
+{
+    int8_t scanResult;
+    wl_status_t status;
+
+    // Fast connect to previous WiFi on startup
+    if (_firstRun) {
+        _firstRun = false;
+
+        // Check if previous WiFi connection saved
+        if (strlen(WiFi.SSID().c_str())) {
+            DEBUG_WIFI_MULTI("[WIFIM] Connecting saved WiFi\n");
+
+            // Connect to previous saved WiFi
+            WiFi.begin();
+
+            // Wait for status change
+            status = waitWiFiConnect(connectTimeoutMs);
+        }
+    }
+
+    // Check connection state
+    status = WiFi.status();
+    if (status == WL_CONNECTED) {
+        // Already connected
+        return status;
+    }
+
+    // Start WiFi scan
+    scanResult = startScan();
+    if (scanResult < 0) {
+        // No WiFi scan results
+        return WL_NO_SSID_AVAIL;
+    }
+
+    // Try to connect to multiple WiFi's with strongest signal (RSSI)
+    return connectWiFiMulti(connectTimeoutMs);
+}
+
+/**
+ * @brief Wait for WiFi connect status change, protected with timeout
+ * @param connectTimeoutMs
+ *      WiFi connection timeout in ms
+ * @return
+ *      WiFi connection status
+ */
+wl_status_t ESP8266WiFiMulti::waitWiFiConnect(uint32_t connectTimeoutMs)
+{
+    wl_status_t status;
+    uint32_t tStart;
+
+    // Wait for status change
+    tStart = millis();
+    do {
+        // Refresh watchdog
+        delay(0);
+
+        // Get WiFi status
+        status = WiFi.status();
+
+        // Check status
+        if (status == WL_CONNECTED) {
+            // Connected, print WiFi status
+            printWiFiStatus(status);
+
+            // Return WiFi status
+            return status;
+        } else if (status == WL_CONNECT_FAILED) {
+            DEBUG_WIFI_MULTI("[WIFIM] Connect failed\n");
+
+            // Return WiFi connect failed
+            return WL_CONNECT_FAILED;
+        }
+    } while ((millis() - tStart) < connectTimeoutMs);
+
+    DEBUG_WIFI_MULTI("[WIFIM] Connect timeout\n");
+
+    return WL_CONNECT_FAILED;
+}
+
+/**
+ * @brief Start WiFi scan
+ * @retval >0
+ *      Number of detected WiFi SSID's
+ * @retval 0
+ *      No WiFi connections found
+ * @retval -2
+ *      WiFi scan failed
+ */
+int8_t ESP8266WiFiMulti::startScan()
```

Comment:
This method doesn't depend on anything from ESP8266WiFiMulti. Please remove it from the class and convert it to a standalone function that is private to the cpp (static or inside anonymous namespace)

Reply:
Accepted.